### PR TITLE
MULE-13896: Improve how URIs are created

### PR DESCRIPTION
**Pull request by mbuchwald**

### CODE REVIEW

File: src/main/java/org/mule/service/http/impl/service/server/ErrorRequestHandler.java [hunk 1]
```diff
@@ -37,7 +38,12 @@ public ErrorRequestHandler(int statusCode, String reasonPhrase, String entityFor
 
   @Override
   public void handleRequest(HttpRequestContext requestContext, HttpResponseReadyCallback responseCallback) {
-    String resolvedEntity = getResolvedEntity(requestContext.getRequest().getUri());
+    String resolvedEntity;
+    try {
+      resolvedEntity = getResolvedEntity(requestContext.getRequest().getUri().toString());
+    } catch (IllegalArgumentException e) {
```
Comment (elrodro83) [inline src/main/java/org/mule/service/http/impl/service/server/ErrorRequestHandler.java]: why is this needed?

File: src/test/java/org/mule/service/http/impl/functional/server/HttpServerErrorRequestsTestCase.java [hunk 2]
```diff
@@ -73,12 +73,14 @@ public void tearDown() {
 
   @Test
   public void methodNotAllowed() throws Exception {
-    verifyErrorResponse(new HttpPost(), "test", METHOD_NOT_ALLOWED, "Method not allowed for endpoint: /test");
+    verifyErrorResponse(new HttpPost(), "test", METHOD_NOT_ALLOWED,
+                        format("Method not allowed for endpoint: http://localhost:%s/test", port.getValue()));
```
Comment (elrodro83) [inline src/test/java/org/mule/service/http/impl/functional/server/HttpServerErrorRequestsTestCase.java]: for generating the error message, the extra generatiod information should not be there to avoid confusion

File: src/test/java/org/mule/service/http/impl/service/server/NoListenerRequestHandlerTestCase.java [hunk 3]
```diff
@@ -29,18 +30,20 @@
...
+  private final static String TEST_REQUEST_INVALID_URI = "http://localhost:8081";
   private final HttpRequestContext context = mock(HttpRequestContext.class, RETURNS_DEEP_STUBS);
   private final HttpResponseReadyCallback responseReadyCallback = mock(HttpResponseReadyCallback.class);
   private NoListenerRequestHandler noListenerRequestHandler;
 
   @Before
   public void setUp() throws Exception {
     noListenerRequestHandler = NoListenerRequestHandler.getInstance();
-    when(context.getRequest().getUri()).thenReturn(TEST_REQUEST_INVALID_URI);
+    when(context.getRequest().getUri()).then(obj -> URI.create(TEST_REQUEST_INVALID_URI));
```
Comment (elrodro83) [inline src/test/java/org/mule/service/http/impl/service/server/NoListenerRequestHandlerTestCase.java]: why not use the cache here?

File: src/main/java/org/mule/service/http/impl/service/server/ErrorRequestHandler.java (hunk 1 above)
Comment (afelisatti) [reply src/main/java/org/mule/service/http/impl/service/server/ErrorRequestHandler.java]: The change in tests suggests `getPath()` should be used here. We don't actually want to expose anything else.

File: src/main/java/org/mule/service/http/impl/service/client/GrizzlyHttpClient.java [hunk 4]
```diff
@@ -397,7 +397,7 @@ private Request createGrizzlyRequest(HttpRequest request, int responseTimeout, b
       builder.setRequestTimeout(responseTimeout);
     });
 
-    reqBuilder.setUrl(request.getUri());
+    reqBuilder.setUrl(request.getUri().toString());
```
Comment (elrodro83) [inline src/main/java/org/mule/service/http/impl/service/client/GrizzlyHttpClient.java]: call setUri

File: src/main/java/org/mule/service/http/impl/service/server/ErrorRequestHandler.java (hunk 1 above)
Comment (mbuchwald) [reply src/main/java/org/mule/service/http/impl/service/server/ErrorRequestHandler.java]: In case of invalid URI, trying to resolve it will result in an `IllegalArgumentException`, making the ErrorRequestHandler not handle the exception

File: src/main/java/org/mule/service/http/impl/service/server/ErrorRequestHandler.java [hunk 5]
```diff
@@ -37,7 +39,14 @@ public ErrorRequestHandler(int statusCode, String reasonPhrase, String entityFor
 
   @Override
   public void handleRequest(HttpRequestContext requestContext, HttpResponseReadyCallback responseCallback) {
-    String resolvedEntity = getResolvedEntity(requestContext.getRequest().getUri());
+    String resolvedEntity;
+    try {
+      URI uri = requestContext.getRequest().getUri();
+      String query = uri.getQuery() != null ? "?" + uri.getQuery() : "";
+      resolvedEntity = getResolvedEntity(uri.getPath() + query);
+    } catch (IllegalArgumentException e) {
```
Comment (elrodro83) [inline src/main/java/org/mule/service/http/impl/service/server/ErrorRequestHandler.java]: still don't know why this is needed. Add a comment

File: src/main/java/org/mule/service/http/impl/service/server/ErrorRequestHandler.java [hunk 6]
```diff
@@ -37,7 +39,14 @@ public ErrorRequestHandler(int statusCode, String reasonPhrase, String entityFor
 
   @Override
   public void handleRequest(HttpRequestContext requestContext, HttpResponseReadyCallback responseCallback) {
-    String resolvedEntity = getResolvedEntity(requestContext.getRequest().getUri());
+    String resolvedEntity;
```
Comment (afelisatti) [inline src/main/java/org/mule/service/http/impl/service/server/ErrorRequestHandler.java]: This is only to indicate that the requested path does not exist or is unavailable, you don't need to expose anything more than that.
